src/sinks/aws_s3/sink.rs

```diff
@@ -0,0 +1,521 @@
+                //
+                // TODO: this might be where we would insert something like the
+                // proposed error handling/dead letter queue stuff; events that
+                // _we_ can't handle, but some other component may be able to
+                // salvage
+                None => {
+                    continue;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+async fn run_io<S>(mut rx: Receiver<S3Request>, mut service: S, acker: Acker)
+where
+    S: Service<S3Request>,
+    S::Future: Send + 'static,
+    S::Response: Response + Send + 'static,
+    S::Error: Debug + Into<crate::Error> + Send,
+{
+    let in_flight = FuturesUnordered::new();
+    let mut pending_acks = HashMap::new();
+    let mut seq_head: u64 = 0;
+    let mut seq_tail: u64 = 0;
+
+    pin!(in_flight);
+
+    loop {
+        select! {
+            Some(req) = rx.recv() => {
+                // Rebind the variable to avoid a bug with the pattern matching
+                // in `select!`: https://github.com/tokio-rs/tokio/issues/4076
+                let mut req = req;
+                let seqno = seq_head;
+                seq_head += 1;
+
+                let (tx, rx) = oneshot::channel();
+
+                in_flight.push(rx);
+
+                trace!(
+                    message = "Submitting service request.",
+                    in_flight_requests = in_flight.len()
+                );
+                // TODO: This likely need be parameterized, which builds a
+                // stronger case for following through with the comment
+                // mentioned below.
+                let logic = StdServiceLogic::default();
+                // TODO: I'm not entirely happy with how we're smuggling
+                // batch_size/finalizers this far through, from the finished
+                // batch all the way through to the concrete request type...we
+                // lifted this code from `ServiceSink` directly, but we should
+                // probably treat it like `PartitionBatcher` and shove it into a
+                // single, encapsulated type instead.
+                let batch_size = req.batch_size;
+                let finalizers = req.take_finalizers();
+
+                let svc = service.ready().await.expect("should not get error when waiting for svc readiness");
+                let fut = svc.call(req)
+                    .err_into()
+                    .map(move |result| {
+                        logic.update_finalizers(result, finalizers);
+
+                        // If the rx end is dropped we still completed
+                        // the request so this is a weird case that we can
+                        // ignore for now.
+                        let _ = tx.send((seqno, batch_size));
+                    })
+                    .instrument(info_span!("request", request_id = %seqno));
+                tokio::spawn(fut);
```

Comment:
It's a little confusing that this is the "IO task", but it seems like we're spawning all of the actual IO off into independent tasks. I could be missing something, but it makes me wonder if it's worthwhile to have this extra task split out with the accompanying channel hop.

Reply:
That's a good question. Absent any specific measure of this code I can't say, but I'll leave this sub-conversation unresolved so it's obvious for later discussion. 

/cc @tobz

Reply:
So we spawn the HTTP call, yes, but the I/O task is the one responsible for _driving_ the I/O.  It's not I/O in the literaly sense of sending and receiving the bytes over the wire, but tracking the request future and dealing with acknowledging the messages (in order) when the responses are resolved.

I modeled it this way as the start of trying to introduce more concurrency into the process.  The plan in my head was that it also probably makes sense to do encoding/compression in the I/O task.

Reply:
FWIW I have the Datadog logs work staged out so that encoding/compression takes place in a separate part of the pipeline from IO.

Reply:
Which is to say, I think it's a good idea. I'll crib ideas from this PR for that work so we'll get to see how it works out in practice in a different sink this week.